File: Problems/No.1379_Find_a_Corresponding_Node_of_a_Binary_Tree_in_a_Clone_of_That_Tree/find_a_corresponding_node_of_a_binary_tree_in_a_clone_of_that_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* getTargetCopy(TreeNode* original, TreeNode* cloned, TreeNode* target) {
        
        // parameters
        std::queue<TreeNode*> Q, Q_clone;
        TreeNode *curr = NULL, *curr_clone = NULL;

        // start computing
        Q.push(original);
        Q_clone.push(cloned);
        while (!Q.empty()) {

            curr = Q.front();
            Q.pop();
            curr_clone = Q_clone.front();
            Q_clone.pop();
            if (curr == target) return curr_clone;

            if (curr->left) Q.push(curr->left);
            if (curr->right) Q.push(curr->right);

            if (curr_clone->left) Q_clone.push(curr_clone->left);
            if (curr_clone->right) Q_clone.push(curr_clone->right);

        }

        return NULL;

    }
};

File: Problems/No.1379_Find_a_Corresponding_Node_of_a_Binary_Tree_in_a_Clone_of_That_Tree/find_a_corresponding_node_of_a_binary_tree_in_a_clone_of_that_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def getTargetCopy(self, original, cloned, target):
        """
        :type original: TreeNode
        :type cloned: TreeNode
        :type target: TreeNode
        :rtype: TreeNode
        """

        # parameters
        Q, Q_clone = [original], [cloned]
        curr, curr_clone = None, None

        # start computing
        while Q:

            curr, curr_clone = Q.pop(), Q_clone.pop()
            if curr is target:
                return curr_clone
            
            if curr.left:
                Q.append(curr.left)
            if curr.right:
                Q.append(curr.right)
            
            if curr_clone.left:
                Q_clone.append(curr_clone.left)
            if curr_clone.right:
                Q_clone.append(curr_clone.right)
        
        return None
